### ISSUE TITLE
Network error while attempting to start a subscription to a Netflix DGS GraphQL Java Server from an Apollo Kotlin Java client

### ISSUE THREAD

**Issue by aviewfromspace1** [apollographql/apollo-kotlin#3991]
**Summary**
I get a generic network error from the Apollo Kotlin library when using it in a Java client to start a subscription to a Netflix DGS Java GraphQL server.

Exception trace returned by Apollo:

```
Network error while executing Alerts
com.apollographql.apollo3.network.ws.WebSocketNetworkTransport$execute$$inlined$map$1$2.emit(Collect.kt:144)
kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:77)
kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:59)
com.apollographql.apollo3.network.ws.WebSocketNetworkTransport$execute$3.invokeSuspend(WebSocketNetworkTransport.kt:249)
com.apollographql.apollo3.network.ws.WebSocketNetworkTransport$execute$3.invoke(WebSocketNetworkTransport.kt)
com.apollographql.apollo3.network.ws.WebSocketNetworkTransport$execute$3.invoke(WebSocketNetworkTransport.kt)
com.apollographql.apollo3.internal.FlowsKt$transformWhile$1$invokeSuspend$$inlined$collectWhile$1.emit(flows.kt:53)
com.apollographql.apollo3.internal.FlowsKt$transformWhile$1$invokeSuspend$$inlined$collectWhile$2.emit(Collect.kt:135)
com.apollographql.apollo3.network.ws.WebSocketNetworkTransport$execute$$inlined$filter$1$2.emit(Collect.kt:137)
kotlinx.coroutines.flow.SubscribedFlowCollector.emit(Share.kt)
kotlinx.coroutines.flow.SharedFlowImpl.collect(SharedFlow.kt:351)
kotlinx.coroutines.flow.SharedFlowImpl$collect$1.invokeSuspend(SharedFlow.kt)
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

**Version**
Apollo Kotlin 3.2.0

**Description**
* Queries and mutations work fine to the server, it is just subscriptions that don't work.
* If I use a Javascript Playground client to start the subscription to the server then it works fine with no errors and I get the expected alert message pushed to the client.

I've got Spring Boot demo app code below that reproduces the error when the defined integration test is run. The code is also attached as a zip file. Please can you take a look and let me know if I am doing something wrong, or if there is an issue or bug to resolve in the Apollo Kotlin client?

Thanks

Matt

**ExampleApplication.java**
```
package com.example;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class ExampleApplication {
public static void main(String[] args) {
	    SpringApplication.run(ExampleApplication.class, args);
	}
}
```

**AlertSubscription.java:**
```
package com.example;

import com.netflix.graphql.dgs.DgsComponent;
import com.netflix.graphql.dgs.DgsSubscription;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Flux;

@DgsComponent
public class AlertSubscription {
    @DgsSubscription
    public Publisher<AlertMessage> alerts() {
        return Flux.just(new AlertMessage("An alert message"));
    }
}
```

**AlertMessage.java:**
```
package com.example;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.ToString;

@AllArgsConstructor
@Getter
@ToString
public class AlertMessage {
    private final String alert;
}
```

**AlertSubscriptionIntegrationTest.java:**
```
package com.example;

import com.apollographql.apollo3.ApolloClient;
import com.apollographql.apollo3.network.ws.SubscriptionWsProtocol;
import com.example.helper.AlertMessageSubscriber;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
class AlertSubscriptionIntegrationTest {

	private ApolloClient apolloClient;

	@BeforeEach
	private void setup() {
		createApolloClient();
	}

	@AfterEach
	private void cleanup() {
		destroyApolloClient();
	}

	private void createApolloClient() {
		if(apolloClient == null) {
			apolloClient = new ApolloClient.Builder()
					.serverUrl("http://localhost:8080/graphql")
					.wsProtocol(new SubscriptionWsProtocol.Factory())
					.webSocketServerUrl("ws://localhost:8080/subscriptions")
					.build();
		}
	}

	private void destroyApolloClient() {
		if(apolloClient != null) {
			try {
				apolloClient.dispose();
			}
			finally {
				apolloClient = null;
			}
		}
	}

	@Test
	protected void receiveMessageOnAlertsSubscription() {
		var expectedAlertMessage = new AlertMessage("An alert message");
		var alertMessageSubscriber = AlertMessageSubscriber.subscribe(apolloClient);
		var receivedAlertMessage = alertMessageSubscriber.waitForMessage(3);
		assertNotNull(receivedAlertMessage, "No message received");
		assertNotNull(receivedAlertMessage.data, "Message contained no payload");
		assertEquals(expectedAlertMessage.getAlert(), receivedAlertMessage.data.alerts.alert, "Didn't receive the expected message");
		alertMessageSubscriber.cancelSubscription();
	}
}
```

**AlertMessageSubscriber.java:**
```
package com.example.helper;

import com.apollographql.apollo3.ApolloCall;
import com.apollographql.apollo3.ApolloClient;
import com.apollographql.apollo3.api.ApolloResponse;
import com.apollographql.apollo3.rx2.Rx2Apollo;
import com.example.client.AlertsSubscription;
import io.reactivex.Flowable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.TimeUnit;

public class AlertMessageSubscriber<T> {
    private Disposable subscriptionRef;
    private final ArrayBlockingQueue<T> receivedMessageQueue = new ArrayBlockingQueue<>(10);

    private AlertMessageSubscriber(){}

    private final Consumer<? super T> onMessageHandler = receivedMessageQueue::add;
    private final Consumer<? super Throwable> onErrorHandler =
            errorMessage -> {
                System.err.println(errorMessage.getMessage());
                for(var traceElement : errorMessage.getStackTrace()) {
                    System.err.println(traceElement);
                }
            };

    public T waitForMessage(final int secondsToWait) {
        T message = null;
        try {
            message = receivedMessageQueue.poll(secondsToWait, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            System.err.println("Wait for message interrupted.");
        }
        return message;
    }

    public static AlertMessageSubscriber<ApolloResponse<AlertsSubscription.Data>> subscribe(final ApolloClient apolloClient) {
        ApolloCall<AlertsSubscription.Data> subscriptionCall = apolloClient.subscription(new AlertsSubscription());
        Flowable<ApolloResponse<AlertsSubscription.Data>> subscriptionResponse = Rx2Apollo.flowable(subscriptionCall);

        var alertMessageSubscriber = new AlertMessageSubscriber<ApolloResponse<AlertsSubscription.Data>>();
        alertMessageSubscriber.subscriptionRef = subscriptionResponse.subscribe(
                alertMessageSubscriber.onMessageHandler,
                alertMessageSubscriber.onErrorHandler
        );

        return alertMessageSubscriber;
    }

    public void cancelSubscription() {
        if(subscriptionRef != null) {
            subscriptionRef.dispose();
            subscriptionRef = null;
        }
    }
}
```

**main/resources/schema/schema.graphql:**
```
type Query {
    alerts: AlertMessage
}

type Subscription {
    alerts: AlertMessage
}

type AlertMessage {
    alert: String!
}
```

**main/resources/schema/alerts.graphql:**
```
subscription Alerts {
    alerts {
        alert
    }
}
```

**main/resources/application/properties:**
```
dgs.graphql.schema-locations=classpath*:schema/**/*.graphqls
```

**build.gradle:**
```
buildscript {
	ext {
		apolloVersion = '3.2.0'
		dgsVersion = '4.9.24'
		lombokVersion = '1.18.22'
		okHttpVersion = '4.9.3'
		springBootVersion = '2.6.6'
		springDependencyMgtPluginVersion = '1.0.11.RELEASE'
	}
}

plugins {
	id 'org.springframework.boot' version "${springBootVersion}"
	id 'io.spring.dependency-management' version "${springDependencyMgtPluginVersion}"
	id 'java'
	id 'com.apollographql.apollo3' version "${apolloVersion}"
}

group = 'com.example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '11'

repositories {
	mavenCentral()
}

dependencies {
	// Spring Boot Dependencies
	implementation(platform("org.springframework.boot:spring-boot-dependencies:${springBootVersion}"))
	implementation "org.springframework.boot:spring-boot-starter-web:${springBootVersion}"
	testImplementation "org.springframework.boot:spring-boot-starter-test:${springBootVersion}"

	// Domain Graph Service (DGS) Dependencies
	implementation(platform("com.netflix.graphql.dgs:graphql-dgs-platform-dependencies:${dgsVersion}"))
	implementation "com.netflix.graphql.dgs:graphql-dgs-spring-boot-starter:${dgsVersion}"
	implementation "com.netflix.graphql.dgs:graphql-dgs-subscriptions-websockets-autoconfigure:${dgsVersion}"

	// Apollo Kotlin library for integration tests
	implementation("com.apollographql.apollo3:apollo-runtime:${apolloVersion}")
	implementation "com.apollographql.apollo3:apollo-rx2-support:${apolloVersion}"
	implementation "com.squareup.okhttp3:okhttp:${okHttpVersion}"

	// lombok
	compileOnly "org.projectlombok:lombok:${lombokVersion}"
	testCompileOnly "org.projectlombok:lombok:${lombokVersion}"
	annotationProcessor "org.projectlombok:lombok:${lombokVersion}"
	testAnnotationProcessor "org.projectlombok:lombok:${lombokVersion}"

	// Addition graphql playground at http://localhost:8080/playground url that supports subscriptions
	implementation group: 'com.graphql-java-kickstart', name: 'playground-spring-boot-starter', version: '11.1.0'
}

test {
	useJUnitPlatform()
}

apollo {
	packageName.set("com.example.client")
	sourceFolder.set("../../../src/main/resources/schema")
}
```

[ExampleApplication.zip](https://github.com/apollographql/apollo-kotlin/files/8413347/ExampleApplication.zip)

**Comment by BoD** [comment]
Hi! Thank you for reporting this, and for the nice repro project 🙏 

I believe this is the same issue as #3990: `graphql-java-kickstart` seems to be expecting text payloads, whereas Apollo Kotlin's `SubscriptionWsProtocol` is sending binary ones. 

I'm not sure about DGS though, and wondering if this could be configurable on that side?

**Comment by BoD** [comment]
[v3.2.1](https://github.com/apollographql/apollo-kotlin/releases/tag/v3.2.1) is now available, and now uses text frames by default. Don't hesitate to re-open or let us know if that resolves the issue.

**Comment by aviewfromspace1** [comment]
I've retested and the issue is fixed in 3.2.1. Awesome effort, thank you @BoD!
